app/models/spree/payment.rb

```diff
@@ -32,7 +32,19 @@ class Payment < ActiveRecord::Base
     # invalidate previously entered payments
     after_create :invalidate_old_payments
 
+    # Prevents #validate_source to end up in a chain of validations, from
```

Comment:
I found this comment quite complicated. Let me try:

First I'd describe what the attribute does and then say where it is used and why:

```ruby
# Skips the validation of the source (for example, credit card) of the payment.
# 
# This is used in refunds as the validation of the card can fail but the refund can go through, we trust the payment 
#    gateway in these cases. For example, Stripe is accepting refunds with source cards that were valid when
#    the payment was placed but are now expired, and we consider them invalid.

```
I think that's it here.

I dont think the chain of validations is related to this refund/skip_source_validation, is it? I think it's related to the invalidation of previous payments that you have wisely changed below from invalidate! to update_column :+1: 
There (line 191), I'd add this comment:
```ruby
# Using update_column skips validations and so it skips validate_source. As we are just invalidating
#     past payments here, we don't want to validate all of them at this stage.
```

Reply:
Thanks, that looks better. 

> I dont think the chain of validations is related to this refund/skip_source_validation, is it?

It was happening due to both. Every source was validating its source up until the credit card.